WIP: Suggestions for various changes

DonatJR opened the pull request:
Some notes regarding these changes:
* The Timer added in `Lcd.cs` is untested as I do not have any hardware to test it on. It would probably be best to check out my fork and run / test changes in the `suggestions branch` before accepting this PR.
* The interval at which this timer writes timestamps to the lcd could be made configurable in the config

Some additional suggestion / things to discuss not implemented in this PR:
1. The [ClientConfig](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Client/ClientConfig.cs#L31) and [ServerConfig](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Server/ServerConfig.cs#L21) gives users the option to change / influence which plugin assemblies are loaded. Maybe this value should be internal only to prevent loading of unauthorized / harmful assemblies. 
1. [The generic PluginType](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Common/PluginLoader.cs#L14) has no type restrictions. You are using it `DataSource` and `DataSink` but it could potentially be used with any class, couldn't it? Would it cause issues / blow up if used if anything but `DataSource` or `DataSink`? If yes maybe there should be a type restriction:
    ```C#
    public static class PluginLoader<PluginType> where PluginType : [base interface of DataSink / 
    DataSource]
    ```
1. What happens if `InitDataSinkPlugins()` [here](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Client/PluginHandler.cs#L212) fails? Do the already initialized sensor plugins need to be shutdown / unitialized?
1. [The dictionary](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Plugins/DataSink/MQTT/MQTT.cs#L32) + [locker object](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Plugins/DataSink/MQTT/MQTT.cs#L37) could probably be replaced with a [ConcurrentDictionary](https://msdn.microsoft.com/en-us/library/dd287191(v=vs.110).aspx) using `TryAdd(...)` & `TryRemove(...)`
1. The property [here](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Client/PluginHandler.cs#L186) could be set inside of the `Init()` method (same for the [server](https://github.com/DonatJR/Weatherstation/blob/24320b6d82e18ea8d480863cae44885a1f9a5e02/Server/PluginHandler.cs#L49))
1. Is the `Environment.Exit(0);` at the end of both the server and client main method necessary? 
1. Config sortorder could be made 1-based instead of 0-based (easier to handle for non-developers)
1. You could add some post-build events to the VS solution to copy built assemblies into the same folder structure as the buildscript does

I will also add some additional comments on files directly.

@ThirtySomething I can't add you as reviewer / assign you as I don't have any rights in your repo. ;)

DonatJR commented on Common/ConfigLoader.cs:
```diff
@@ -76,9 +75,9 @@ protected static ConfigClass LoadFromFile(string fileName)
             using (var reader = new StreamReader(FileName))
             {
+                var serializer = new XmlSerializer(typeof(ConfigClass));
                 return (ConfigClass)serializer.Deserialize(reader);
```
this cast could throw, maybe this should be handled?

DonatJR replied:
you are right, the `reader` could throw as well, but:
what if you are reading a file with invalid contents? either the `serializer` will throw when trying to deserialize it or it will accept the invalid content silently (not sure about the `XmlSerializer` class) and the [cast will throw ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/casting-and-type-conversions#type-conversion-exceptions-at-run-time). either way it never hurts to be extra careful when doing casts ...

DonatJR commented on Plugins/DataSink/SQLite/SQLite.cs:
```diff
@@ -107,8 +105,8 @@ private void InsertData(MeasurementValue sensorValue)
             string stm = $"INSERT INTO {sensorValue.Name} (timestamp, value, hash) VALUES (@timestamp, @value, @hash)";
```
the unsanitized `sensorValue.Name` could potentionally be an attack vector for sql injection (this is not the only occurrence in this file, there are a few other statements with the same problem)

ThirtySomething replied:
Correct - I'm thinking about a different ERM. One table for measurement value, one table for units, one table for the source. Then all tablenames are fixed - this will make the implementation of the Web part more easier. And it will fix the injection problem.

DonatJR replied:
I'd be happy to take a look at it once you changed the ERM / have a concrete proposal :+1:

ThirtySomething commented on Plugins/DataSink/Lcd/LcdConfig.cs:
```diff
@@ -15,6 +15,6 @@ public class LcdConfig : ConfigLoader<LcdConfig>, IConfigSaver
         /// Format to display a timestamp
         /// </summary>
-        public string TimestampFormat = "dd.MM.yyyy  HH:mm:ss";
+        public string TimestampFormat = "dd.MM.yyyy HH:mm:ss";
```
This space was placed with reason there. The LCD has a width of 20 digits. In this mask the date is left bound, the time is right bound.

DonatJR replied:
I thought it was a typi, will change it back :+1:

ThirtySomething commented on Plugins/DataSink/Lcd/Readme.md:
```diff
@@ -10,4 +10,4 @@ For the lcd plugin the following settings are available:
 
 * <code>BrickDaemonIP</code> - The IP of the system where the Tinkerforge brick daemon is running. Usually the same system where the weather station is connected to: <code>localhost</code>
 * <code>BrickDaemonPort</code> - The port of the brick daemon, the default is <code>4223</code>
-* <code>TimestampFormat</code> - The format of the timestamp printed to the lcd display, the default is <code>dd.MM.yyyy  HH:mm:ss</code>
```
See remark `Plugins/DataSink/Lcd/LcdConfig.cs`

ThirtySomething commented on Plugins/DataSink/MQTT/MQTTConfig.cs:
```diff
@@ -8,7 +8,7 @@ public class MQTTConfig : ConfigLoader<MQTTConfig>, IConfigSaver
         /// IP of MQTT broker to connect to
         /// </summary>
-        public string BrokerIP { get; set; } = "127.0.0.1";
+        public string BrokerIP { get; set; } = "localhost";
```
`127.0.0.1` will be always available. You can modify the `localhost` entry in the host file to point to somewhere else.

DonatJR replied:
on the other hand localhost is protocol independent (works for IPv4 and IPv6). and there are a few occasions of localhost already in the code.
tell me which one you prefer and I will change it everywhere :)

ThirtySomething commented on Plugins/DataSink/MQTT/Readme.md:
```diff
@@ -1,14 +1,14 @@
 For the MQTT plugin the following settings are available:
 
-* <code>BrokerIP</code> - The IP of the system where the MQTT broker is running. The default is: <code>127.0.0.1</code>
+* <code>BrokerIP</code> - The IP of the system where the MQTT broker is running. The default is: <code>localhost</code>
```
`127.0.0.1` will be always available. You can modify the `localhost` entry in the host file to point to somewhere else.

DonatJR replied:
see the other comment for a reply

ThirtySomething commented on Server/Readme.md:
```diff
@@ -13,10 +13,10 @@ Return to [main](./../Readme.md).
+* <code>Delay</code> - time interval at which the next keypress is checked
+* <code>PluginPath</code> - The path where the client is looking for datasource and datasink plugins, by default in a subdirectory called <code>Plugins</code>
+* <code>PluginProductName</code> - Each plugin should have a specified product name to identify them as plugins for this software and to ignore non-plugin DLL files. The default is <code>net.derpaul.tf.plugin.[name]</code> and this is valid __FOR ALL__ plugins.
+* <code>BrokerIP</code> - The IP of the system where the MQTT broker is running. The default is: <code>localhost</code>
```
`127.0.0.1` will be always available. You can modify the `localhost` entry in the host file to point to somewhere else.

DonatJR replied:
see the other comment for a reply

ThirtySomething commented on Server/ServerConfig.cs:
```diff
@@ -23,7 +23,7 @@ public class ServerConfig : ConfigLoader<ServerConfig>, IConfigSaver
         /// IP of MQTT broker to connect to
         /// </summary>
-        public string BrokerIP { get; set; } = "127.0.0.1";
+        public string BrokerIP { get; set; } = "localhost";
```
`127.0.0.1` will be always available. You can modify the `localhost` entry in the host file to point to somewhere else.

DonatJR replied:
see the other comment for a reply

ThirtySomething commented on Plugins/DataSource/Barometer/Barometer.cs:
```diff
@@ -24,18 +24,16 @@ public abstract class Barometer : SensorBase
         public override void Init(IPConnection connection, string UID)
         {
-            if (_Bricklet != null)
+            if (_Bricklet == null)
```
To have a common handling, maybe `_Bricklet` should be named also `Bricklet` like in the other `IDataSource` plugins?

ThirtySomething commented on Server/Server.cs:
```diff
@@ -61,13 +61,17 @@ private void Run()
-                    while (!System.Console.KeyAvailable)
+                    if (!System.Console.KeyAvailable)
+                    {
+                        System.Threading.Thread.Sleep(ServerConfig.Instance.Delay);
```
Same question like on the `Client` - I'd prefer to have a single method call in case the thread sleep will be replaced with a better solution. This will not bloat the loop.

DonatJR replied:
I replied on the other comment ;)

ThirtySomething commented on Plugins/DataSource/Barometer/Temperature.cs:
```diff
@@ -14,18 +14,16 @@ public class Temperature : Barometer
         {
             MeasurementValue result = new MeasurementValue(Name, Unit, TemperatureConfig.Instance.SortOrder);
 
             if (_Bricklet == null)
```
This should check `_Bricklet != null` - otherwise the temperature will never be read from the sensor.

DonatJR replied:
thanks for catching that!

ThirtySomething commented on Plugins/DataSink/Lcd/Lcd.cs:
```diff
@@ -58,35 +70,56 @@ public void Shutdown()
+                    lock (this.Locker)
+                    {
+                        // Display timestamp
+                        Bricklet.WriteLine(0, 0, DateTime.Now.ToString(LcdConfig.Instance.TimestampFormat));
```
Additional I discovered an asynchronous behaviour - the values are updated different from the timestamp :-(

DonatJR replied:
First point: You updated the timestamp with every new value (be it temperature, lux or something else) so only the timestamp of the last value recorded was shown on the lcd. I don't think information about when a value was read from a sensor is of any interest to someone reading the values from the lcd, especially if it is just one of several different timestamps

Second point: I thought the intention was to only show a timestamp when it has a clean value (i.e. seconds divisible by 5). This change achieves exactly that and also updates values as soon as they are available, but I get that my assumption was incorrect.

What do you think about building a buffer for all sensor values (only saving the last value from each sensor) and then updating the lcd with information and the time of update (i.e. when the values on the lcd were last refreshed) from this buffer every 5 seconds?

ThirtySomething replied:
The idea of the buffer seems to be a good one - on the first look. As I started with this project, I've had exactly this: A collection of all values recorded in one step. But I ran into trouble with the handshake process between the MQTT plugin of the client and the server. So I decided to publish each sensor value without the buffer. If you want to publish a buffer, this is a big change. Basically each data sink should be able to deal with this buffer. And you need also to change the PluginHandler::HandleValues method of the client, too. Feel free to optimize this ;-)

ThirtySomething commented on Common/TFUtils.cs:
```diff
@@ -8,25 +8,17 @@ namespace net.derpaul.tf
                 currentSeconds = DateTime.Now.Second;
             }
         }
-
```
Really nice method - works much better than the original one :+1:

ThirtySomething commented on Common/TFUtils.cs:
```diff
@@ -8,25 +8,17 @@ namespace net.derpaul.tf
-        /// Wait until n milliseconds are gone
-        /// </summary>
-        /// <param name="delay"></param>
-        public static void WaitNMilliseconds(int delay)
```
See the remark on the `Client` and/or on the `Server`.

DonatJR replied:
replied on the first comment

ThirtySomething commented:
About your suggestions/proposals:

* Personally I really irritated the timer in the LCD plugin for two reasons. First the timestamp of the sensor value is now ignored and instead the system time is taken. Second the timer updates not synchron to the measurement values. I would prefer the original version here without the timer.

1. You're right, the setting of the topic in a config file might be harmful. But... how will you make it configurable on the one hand - and on the other hand keep it save and secure?
2. Yes, the generic PluginType was defined in a way for common usage. Maybe in other projects, so I prefer to keep it universal here.
3. Well, the data source plugins do not really need a shutdown. So the interface does not offer a shutdown method. The opposite are the data sink plugins - they have internal resources needing a controlled shutdown. Flushing a file, closing a database, in case of the LCD poweroff the background light, ...
4. The ConcurrentDictionary is a good option because it might be faster and allows concurrent access. Just want to learn something about locking here. 
5. Yes, you're right. Do you want to change this also? Or do you prefer to open an issue?
6. I'm not so experienced with C# to know this. But I've discovered some strange behaviour until I've added these lines. Maybe I have to investigate a little bit more how to perform a clean exit.
7. Yes and no. Basically you're right, but if you change the sort order, you have to change also the way the position is calculated in the LCD plugin. Could you please take care about this? I've seen somewhere a position incrementation, but not sure if this is valid for all.
8. Hm, I've had this in the past. But unfortunately this means manual modifying of the *.csproj files. Additional the project build of client/server will fail if one of the plugins have not been able to build. Any proposals to solve this scenario?

Followed your invitation and added myself as reviewer :D

DonatJR commented on Client/Client.cs:
```diff
@@ -16,20 +16,23 @@ private static void Main()
                 {
                     pluginHandler.HandleValues(pluginHandler.ValuesRead());
-                    TFUtils.WaitNMilliseconds(ClientConfig.Instance.Delay);
+                    System.Threading.Thread.Sleep(ClientConfig.Instance.Delay);
```
I think sleeping is fine here, why do you think it is not satisfactory?
(regardless: Once there is an improvement it should of course be its own function but until then I would avoid another layer of indirection which has no real value / no code if its own)

DonatJR commented on Common/PluginLoader.cs:
```diff
@@ -68,6 +67,7 @@ private static List<Type> GetPluginTypeList(List<Assembly> assemblyList)
             {
                 return null;
```
good call (also, avoiding null is always an improvement :))

DonatJR commented:
Alright, I made almost all of your requested changes. The only remaining point is the update of the LCD screen, I added a comment on how to resolve this in April. Do you want me to go ahead with my suggestion ([using a buffer](https://github.com/ThirtySomething/Weatherstation/pull/1#discussion_r179436484)), revert the relevant code completely or think about another solution (I could create an issue for this)?

Comments for the general points:
__1:__ I would not make it configurable at all and keep it complete in the code. Using plugin assemblies from another source should be _carefully_ considered and in my opinion merits changes in the code (you could have an internal list with sources you trust and only allow adding plugins from these sources).

__5:__ I changed my mind on this as it would require to set this flag for every class implementing `IDataSink`, the current solution is more elegant.

__6:__ According to this [StackOverflow question](https://stackoverflow.com/a/692371) calling `Environment.Exit()` is only necessary if there are other threads running at the end of `main` as the call terminates them. And even then it would be better to terminate the threads in a controlled manner.

__7:__ I created an issue for this: #3 (you can assign me if you want)

__8:__ I'm not sure why you would need to modify the project files by hand? There is a menu for post-build events in the project properties (these should have no effect on whether or not the build of other projects will fail):
![image](https://user-images.githubusercontent.com/16246147/40283755-a4dbd102-5c83-11e8-8596-b017ab92d15f.png)

ThirtySomething commented:
I've merged your commit(s) and assigned the issue #3 to you. I've also created the issues #4 and #5 according to your proposals and assigned them to myself. For the idea with the buffer related to the LCD first read my comment. If you want to discuss this topic I propose to open an issue just to get a better overview about the discussion.
